{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNA sequence\n",
      "MATVRE\n"
     ]
    }
   ],
   "source": [
    "# open() function is used to open a file before we can read the file; it opens but nothing will be shown; it take one argument; and returns a file objecct (= a file on the computer's hard drive. it is different than strings and/or numbers, where after assigning them to a variable we could use the variable name, e.g., my_string = 'abcdefg'; print(my_string); my_number = 42; print(my_number + 1))\n",
    "# e.g.;\n",
    "#! /usr/bin/env python3\n",
    "\n",
    "my_file = open(\"dna.txt\")\n",
    "\n",
    "#while in the case of file objects we are working (interacting) with them through methods, like read();\n",
    "\n",
    "# read() is a method to read the content of a file object; it does not take any argument, and the return value is a string, which we can store in a variable, then variable name could be used in the subsequent codes;\n",
    "file_contents = my_file.read()\n",
    "print(file_contents)\n",
    "\n",
    "# So, the variable name, to which we assigned the content of the file with a method, like read(), could be use as the variable name of a string in the subsequent codes;\n",
    "\n",
    "# To find the compelent of the DNA in the file, use the following codes;\n",
    "#for i in file_contents:\n",
    "#    if (i == \"A\"): print(\"T\", end = '')\n",
    "#    if (i == \"T\"): print(\"A\", end = '')\n",
    "#    if (i == \"C\"): print(\"G\", end = '')\n",
    "#    if (i == \"G\"): print(\"C\", end = '')\n",
    "\n",
    "# To create a file like txt,\n",
    "#my_file = open(\"dna.txt\", \"w\")\n",
    "#my_file.write(\"\"\"Important\"\"\")\n",
    "#my_file.close()\n",
    "\n",
    "# To append to the reviously cretaed fle lie txt\n",
    "#my_file = open(\"dna.txt\", \"a\")\n",
    "#my_file.write(\"\"\" DNA sequence\"\"\")\n",
    "#my_file.close()\n",
    "\n",
    "# To append to the reviously cretaed fle lie txt\n",
    "#my_file = open(\"dna.txt\", \"a\")\n",
    "#my_file.write(\"\"\"\n",
    "#ACTGTACGTGCACTGATC\"\"\")\n",
    "#my_file.close()\n",
    "\n",
    "# To open the file in python\n",
    "#my_file = open(\"dna.txt\", \"r\")\n",
    "#text=my_file.read()\n",
    "#print(text)\n",
    "#my_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RNA sequence\n",
      "MATVRE\n"
     ]
    }
   ],
   "source": [
    "# my_file_name = \"dna.txt\" is showing a file name (=string).\n",
    "# my_file = open(\"dna.txt\") is showing a file object, i.e., the file is opened, but it is not read, yet.\n",
    "# my_file_contents = my_file.read() is showing a file contents (=string, .e., the text found in the text file).\n",
    "# The variable names are arbitrary. The following codes the same as the one above;\n",
    "apple = \"dna.txt\"\n",
    "banana = open(apple)\n",
    "grape = banana.read()\n",
    "print (grape)\n",
    "\n",
    "# Except it is harder to understand.\n",
    "\n",
    "# But the file name is not constant it should correspond to the name of a file on the hard drive of your computer.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence is RNA sequence\n",
      "MATVRE and the sequence length is 19\n"
     ]
    }
   ],
   "source": [
    "# 3 Dealing with newline\n",
    "\n",
    "# Read the DNA sequence in dna.txt file using the functions and methods we got so far;\n",
    "\n",
    "# Open the file;\n",
    "\n",
    "#my_file = open(\"dna.txt\")\n",
    "\n",
    "# Read the contents;\n",
    "\n",
    "#my_dna = my_file.read()\n",
    "\n",
    "# Calculate the length;\n",
    "\n",
    "#dna_length = len(my_dna)\n",
    "\n",
    "# Print the output;\n",
    "\n",
    "#print(\"sequence is \" + my_dna + \" and the sequence length is \" + str(dna_length))\n",
    "\n",
    "# two mistaks, because of having a new line at the end of the dna sequence;\n",
    "# may consider the new line at the end of the dna sequence as a character, the dana may be ike,\"ACTGTACGTGCACTGATC\\n\"\n",
    "# first mistake, 1) it counted the newline '\\n' at the end as a character; second 2) used the \\n as an order to break down the output onto two lines;\n",
    "\n",
    "# As follow;\n",
    "# sequence is ACTGTACGTGCACTGATC\n",
    "# and the sequence length is 19\n",
    "\n",
    "# while it should be only 18 NTs and show the ouput on one line; \n",
    "\n",
    "## Solution is simple;\n",
    "# This is a common problem with strings. Python settings have a method to remove the newlines from the end of the strings, th method is called rstrip(),\n",
    "# rstrip() is a method which takes one string argument which is the character that you want to remove from the output, here, we want to remove the \\\\n\\ new line.\n",
    "# so the above codes will be updated as follow:\n",
    "\n",
    "# Open the file;\n",
    "\n",
    "my_file = open(\"dna.txt\")\n",
    "\n",
    "\n",
    "# Read the contents;\n",
    "\n",
    "#my_file_contents = my_file.read()\n",
    "\n",
    "# REMOVE the newline from the end of the file contents\n",
    "#my_dna = my_file_contents.rstrip(\"\\n\")\n",
    "\n",
    "\n",
    "# here wa are using two methds read() & rstrip() in the same statement.\n",
    "my_dna = my_file.read().rstrip(\"\\n\")\n",
    "\n",
    "# if did not work or was difficult we can break the statement to two staements as above.\n",
    "\n",
    "\n",
    "# Calculate the length;\n",
    "\n",
    "dna_length = len(my_dna)\n",
    "\n",
    "# Print the output;\n",
    "\n",
    "print(\"sequence is \" + my_dna + \" and the sequence length is \" + str(dna_length))\n",
    "\n",
    "\n",
    "# We could shorten the above two steps as follow:\n",
    "#my_file_contents = my_file.read()\n",
    "\n",
    "#&\n",
    "#my_dna = my_file_contents.rstrip(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'noneexistent.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-d6b75c3f3787>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# 4 Missing files;\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m# If there was no the file you are asking to open, will get an error message saying, \"No such file or directory: 'noneexisttent.txt';\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mmy_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"noneexistent.txt\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;31m# the error is as follow:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# FileNotFoundError: [Errno 2] No such file or directory: 'noneexistent.txt'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'noneexistent.txt'"
     ]
    }
   ],
   "source": [
    "# 4 Missing files;\n",
    "# If there was no the file you are asking to open, will get an error message saying, \"No such file or directory: 'noneexisttent.txt';\n",
    "my_file = open(\"noneexistent.txt\")\n",
    "# the error is as follow:\n",
    "# FileNotFoundError: [Errno 2] No such file or directory: 'noneexistent.txt'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Writing text to files;\n",
    "# So far we added output to screen, it is good if e do not have too much data. \n",
    "# it is not easy to search in output that's being displayed atterminal, and long lines tend to wrapped.\n",
    "# Most important terminal output vanishes when you close the terminal.\n",
    "\n",
    "# Opening files for writing;\n",
    "# As we said we can open files uisng open(), we can also write to files, but we need to use open() in slightly a different way.\n",
    "# two-argument version of open() is used to open a file for writing to it. The second argument (called mode of the file) is the string or the text we want to write in the file.\n",
    "# The second argument could be 'w', 'a', or 'r', for writing appending or reading the file, respectively.\n",
    "# Also, if we did not write anything for the second argument, the default is 'r' by Python.\n",
    "# 'w' and 'a' will create a file if was not exist.\n",
    "\n",
    "\n",
    "# write() method;\n",
    "# write() could also be used to write to a file. it likes print(\"text\") function, takes a srting argumnet, but the difference is i  print it prints to screen, while in write(\"text\") it writes the text to a file instead of showing it on screen.\n",
    "# e.g.,\n",
    "\n",
    "my_file = open(\"out1.txt\", \"w\")\n",
    "my_file.write(\"Hello world! This is a text to be written in a file \\n\\n\")\n",
    "# after running the above codes no txt will get printed to screen, we need to open the file to double check if it is working.\n",
    "\n",
    "# The following examples are all OK;\n",
    "# Write \"abcdef\"\n",
    "my_file.write(\"abc\" + \"def\\n\\n\")\n",
    "\n",
    "# Write \"8\"\n",
    "my_file.write(str(len('AGTGCTAG')) + \"\\n\\n\")\n",
    "\n",
    "# Write \"TTGC\"\n",
    "my_file.write(\"ATGC\".replace('A', 'T') + \"\\n\\n\")\n",
    "\n",
    "# Write \"atgc\"\n",
    "my_file.write(\"ATGC\".lower() + \"\\n\\n\")\n",
    "\n",
    "# Write contents of my variable\n",
    "my_file.write(\"my variable\\n\")\n",
    "\n",
    "my_file.close() # nothing will be passed to the file if the code was not closed.\n",
    "\n",
    "\n",
    "# Closing files; using close() method\n",
    "# This is the opposite of open(). But close() is a method, while open() is a function.\n",
    "# We should call (write) close() after we done with our reading and writing to a file.\n",
    "# Close() is an unsual method as it takes NO arguments (so it's called with an empty pair of parentheses) and does not return any usful value.\n",
    "# e.g.,\n",
    "my_file = open(\"out2.txt\", \"w\")\n",
    "my_file.write(\"Hello World!\")\n",
    "# remember to close the file, as follow:\n",
    "my_file.close()\n",
    "\n",
    "# Paths and folders;\n",
    "# So far we just opened files located in the same directory of the python script where we are running our program, but \n",
    "# To open a file from a different path or directories; open() function can deal with files located in different directories (folders) or baths however you need to provide the full path of the file.\n",
    "\n",
    "#my_file = open(\"/Users/abdulkarimshwani/Desktop/ppfbiologists/Bio_Project/paths_folder.txt\", \"w\")\n",
    "#my_file.write(\"Hello ! This is for paths & folders\")\n",
    "#my_file.close()\n",
    "\n",
    "# e.g.;\n",
    "my_file = open(\"/Users/abdulkarimshwani/Desktop/ppfbiologists/Bio_Project/paths_folder.txt\")\n",
    "\n",
    "# So here the object file is just opened, to see the content, use;\n",
    "#contents = my_file.read() and then print(contents)\n",
    "# or istead of the two line, use one line as follow; \n",
    "#print(my_file.read())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exercises from b-4-bch03;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exercise1; Splititng genomic DNA; All of the following solutions, especially the last one, are working!\n",
    "\n",
    "#my_dna = \"ATCGATCGATCGATCGACTGACTAGTCATAGCTATGCATGTAGCTACTCGATCGATCGATCGATCGATCGATCGATCATGCTATCATCGATCGATATCGATGCATCGACTACTAT\"\n",
    "#exon1 = my_dna[0:63]\n",
    "#print(exon1)\n",
    "#intron = my_dna[63:90]\n",
    "#print(intron.lower())\n",
    "#exon2 = my_dna[90:]\n",
    "#print(exon2)\n",
    "#print(\"The entire DNA sequence is as follow:\" + \"\\n\" + (exon1 + intron.lower() + exon2))\n",
    "#print(exon1 + intron.lower() + exon2)\n",
    "#To add each sequene on a new line:\n",
    "#print(\"The exon1 sequence is as follow:\" + \"\\n\" + exon1 + \"\\n\" + \"The intron sequence is as follow:\" + \"\\n\" + intron.lower() + \"\\n\" + \"The exon2 sequence is as follow:\" + \"\\n\" + exon2 + \"\\n\")\n",
    "\n",
    "# To put each sequence in a file;\n",
    "#my_file = open(\"exon1.txt\", \"w\")\n",
    "#my_file.write(\"#The exon1 sequence is as follow:\" + \"\\n\" + exon1)\n",
    "#my_file = open(\"intron.txt\", \"w\")\n",
    "#my_file.write(\"#The intron sequence is as follow:\" + \"\\n\" + intron.lower())\n",
    "#my_file = open(\"exon2.txt\", \"w\")\n",
    "#my_file.write(\"#The exon2 sequence is as follow:\" + \"\\n\" + exon2)\n",
    "#my_file.close()\n",
    "\n",
    "#from ch03 we need to read from files rather than put the DNA in the code;\n",
    "#my_file = open(\"genomic_dna.txt\")\n",
    "#my_dna = my_file.read()\n",
    "#print(\"The opened DNA:\" + \"\\n\" + my_dna)\n",
    "#exon1 = my_dna[0:63]\n",
    "#intron = my_dna[63:90]\n",
    "#exon2 = my_dna[90:]\n",
    "# To put each sequence in a file;\n",
    "#my_file = open(\"exon1.txt\", \"w\")\n",
    "#my_file.write(\"#The exon1 sequence is as follow:\" + \"\\n\" + exon1)\n",
    "#my_file = open(\"intron.txt\", \"w\")\n",
    "#my_file.write(\"#The intron sequence is as follow:\" + \"\\n\" + intron.lower())\n",
    "#my_file = open(\"exon2.txt\", \"w\")\n",
    "#my_file.write(\"#The exon2 sequence is as follow:\" + \"\\n\" + exon2)\n",
    "#my_file.close()\n",
    "\n",
    "#from ch03 we need to read from files rather than put the DNA in the code;\n",
    "#dna_file = open(\"genomic_dna.txt\")  #  file object opened\n",
    "#my_dna = dna_file.read()  #  file content read\n",
    "#print(\"The read entire DNA equence:\" + \"\\n\" + my_dna) #  file content printed on screen\n",
    "#exon1 = my_dna[0:63]\n",
    "#intron = my_dna[63:90]\n",
    "#exon2 = my_dna[90:]\n",
    "# To put each sequence in a file;\n",
    "#exon1_file = open(\"exon1.txt\", \"w\")\n",
    "#exon1_file.write(\"#The exon1 sequence is as follow:\" + \"\\n\" + exon1)\n",
    "#intron_file = open(\"intron.txt\", \"w\")\n",
    "#intron_file.write(\"#The intron sequence is as follow:\" + \"\\n\" + intron.lower())\n",
    "#exon2_file = open(\"exon2.txt\", \"w\")\n",
    "#exon2_file.write(\"#The exon2 sequence is as follow:\" + \"\\n\" + exon2)\n",
    "#dna_file.close()\n",
    "#exon1_file.close()\n",
    "#intron_file.close()\n",
    "#exon2_file.close()\n",
    "\n",
    "#from ch03 we need to read from files rather than put the DNA in the code;\n",
    "#my_file = open(\"genomic_dna.txt\")  #  file object opened\n",
    "#my_dna = my_file.read()  #  file content read\n",
    "# To put each sequence in a file;\n",
    "#my_file = open(\"exon1.txt\", \"w\")\n",
    "#my_file.write(\"# The exon1 sequence is as follow:\" + \"\\n\" + my_dna[0:63])\n",
    " # Here and below, we cannot put (my_dna[0:63]) in between quotes (\"my_dna[0:63]\"), even though the codes look work.\n",
    "#my_file = open(\"intron.txt\", \"w\")\n",
    "#my_file.write(\"# The intron sequence is as follow:\" + \"\\n\" + my_dna[63:90].lower())\n",
    "#my_file = open(\"exon2.txt\", \"w\")\n",
    "#my_file.write(\"# The exon2 sequence is as follow:\" + \"\\n\" + my_dna[90:])\n",
    "#my_file.close()\n",
    "\n",
    "#To open the dna file and read its contents:\n",
    "dna_file = open(\"genomic_dna.txt\")  #  file object opened\n",
    "my_dna = dna_file.read()  #  file content read\n",
    "\n",
    "# To extract the different bits of DNA sequence:\n",
    "exon1 = my_dna[0:63]\n",
    "intron = my_dna[63:90]\n",
    "exon2 = my_dna[90:]\n",
    "\n",
    "# To open the two output files:\n",
    "coding_file = open(\"coding_dna.txt\", \"w\")\n",
    "noncoding_file = open(\"noncoding_dna.txt\", \"w\")\n",
    "\n",
    "# To write the sequene to the output files:\n",
    "coding_file.write(\"# The coding_sequence sequence is as follow:\" + \"\\n\" + exon1 + exon2)\n",
    "noncoding_file.write(\"# The None_coding_sequence is as follow:\" + \"\\n\" + intron.lower())\n",
    "dna_file.close()\n",
    "coding_file.close()\n",
    "noncoding_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exercise2; Writing a FASTA file: All of the following answers could work!\n",
    "\n",
    "#>sequence_name\n",
    "#ATCGACTGATCGATCGTACGAT # the sequence\n",
    "\n",
    "#Question: Write a program that will create a FASTA file for the following 3 sequences - make sure that all sequences are in uppercase and only contain the bases A, T, G, and C.\n",
    "\n",
    "# Sequence header:               DNA sequence:\n",
    "# ABC123                         ATCGTACGATCGATCGATCGCTAGACGTATCG\n",
    "# DEF456                         actgatcgacgatcgatcgatcacgact\n",
    "# HIJ789                         ACTGAC-ACTGT--ACTGTA----CATGTG\n",
    "\n",
    "#fasta_file = open(\"FASTA.fasta\",\"w\")\n",
    "#fasta_file.write(\">ABC123\\n\" + \"ATCGTACGATCGATCGATCGCTAGACGTATCG\\n\" + \">DEF456\\n\" + \"actgatcgacgatcgatcgatcacgact\\n\" + \">HIJ789\\n\" + \"ACTGAC-ACTGT--ACTGTA----CATGTG\\n\")\n",
    "#my_file.write(\">ABC123\\n\" + \"ATCGTACGATCGATCGATCGCTAGACGTATCG\\n\" + \">DEF456\\n\" + \"actgatcgacgatcgatcgatcacgact\\n\".upper() + \">HIJ789\\n\" + \"ACTGAC-ACTGT--ACTGTA----CATGTG\\n\".rstrip('-')) # rstrip('-') does not work, because it used to remove new lne breaks only!\n",
    "#fasta_file.close()\n",
    "\n",
    "my_file = open(\"FASTA.fasta\",\"w\")\n",
    "my_file.write(\">ABC123\\n\" + \"ATCGTACGATCGATCGATCGCTAGACGTATCG\\n\" + \">DEF456\\n\" + \"actgatcgacgatcgatcgatcacgact\\n\".upper() + \">HIJ789\\n\" + \"ACTGAC-ACTGT--ACTGTA----CATGTG\\n\".replace('-',''))\n",
    "my_file.close()\n",
    "\n",
    "# Answer from the book:\n",
    "\n",
    "header_1 = \"ABC123\"\n",
    "header_2 = \"DEF456\"\n",
    "header_3 = \"HIJ789\"\n",
    "seq_1 = \"ATCGTACGATCGATCGATCGCTAGACGTATCG\"\n",
    "seq_2 = \"actgatcgacgatcgatcgatcacgact\"\n",
    "seq_3 = \"ACTGAC-ACTGT--ACTGTA----CATGTG\"\n",
    "\n",
    "# print them to screen;\n",
    "#print(header_1)\n",
    "#print(seq_1)\n",
    "#print(header_2)\n",
    "#print(seq_2)\n",
    "#print(header_3)\n",
    "#print(seq_3)\n",
    "\n",
    "#print ('>' + header_1 + '\\n' + seq_1)\n",
    "#print ('>' + header_2 + '\\n' + seq_2)\n",
    "#print ('>' + header_3 + '\\n' + seq_3)\n",
    "\n",
    "#print ('>' + header_1 + '\\n' + seq_1)\n",
    "#print ('>' + header_2 + '\\n' + seq_2.upper())\n",
    "#print ('>' + header_3 + '\\n' + seq_3.replace('-',''))\n",
    "\n",
    "# Switch from print to screen t type to files;\n",
    "#output = open(\"sequences.fasta\", \"w\")\n",
    "#output.write('>' + header_1 + '\\n' + seq_1)\n",
    "#output.write('>' + header_2 + '\\n' + seq_2.upper())\n",
    "#output.write('>' + header_3 + '\\n' + seq_3.replace('-',''))\n",
    "#output.close()\n",
    "\n",
    "# put a new line at the end of the DNA sequence;\n",
    "\n",
    "output = open(\"sequences.fasta\", \"w\")\n",
    "output.write('>' + header_1 + '\\n' + seq_1 + '\\n')\n",
    "output.write('>' + header_2 + '\\n' + seq_2.upper() + '\\n')\n",
    "output.write('>' + header_3 + '\\n' + seq_3.replace('-','') + '\\n')\n",
    "output.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exercise3; Writing multiple FASTA files; All the following solutions are ok;\n",
    "\n",
    "#Question: \n",
    "# Sequence header:               DNA sequence:\n",
    "# ABC123                         ATCGTACGATCGATCGATCGCTAGACGTATCG\n",
    "# DEF456                         actgatcgacgatcgatcgatcacgact\n",
    "# HIJ789                         ACTGAC-ACTGT--ACTGTA----CATGTG\n",
    "\n",
    "# My solution;\n",
    "#My_file = open(\"first_fasta.fasta\",\"w\")\n",
    "#My_file.write(\">ABC123\\n\" + \"ATCGTACGATCGATCGATCGCTAGACGTATCG\")\n",
    "#My_file = open(\"second_fasta.fasta\",\"w\")\n",
    "#My_file.write(\">DEF456\\n\" + \"actgatcgacgatcgatcgatcacgact\".upper())\n",
    "#My_file = open(\"third_fasta.fasta\",\"w\")\n",
    "#My_file.write(\">HIJ789\\n\" + \"ACTGAC-ACTGT--ACTGTA----CATGTG\".replace('-',''))\n",
    "#My_file.close()\n",
    "\n",
    "#Solution found from the book;\n",
    "# set the values of all the header variable\n",
    "header_1 = \"ABC123\"\n",
    "header_2 = \"DEF456\"\n",
    "header_3 = \"HIJ789\"\n",
    "\n",
    "#set the values of all the sequence variables\n",
    "seq_1 = \"ATCGTACGATCGATCGATCGCTAGACGTATCG\"\n",
    "seq_2 = \"actgatcgacgatcgatcgatcacgact\"\n",
    "seq_3 = \"ACTGAC-ACTGT--ACTGTA----CATGTG\"\n",
    "\n",
    "# make three files to hold the output\n",
    "output_1 = open(header_1 + \".fasta\",\"w\")\n",
    "output_2 = open(header_2 + \".fasta\",\"w\")\n",
    "output_3 = open(header_3 + \".fasta\",\"w\")\n",
    "\n",
    "# write one sequence to each output file\n",
    "output_1.write('>' + header_1 + '\\n' + seq_1 + '\\n')\n",
    "output_2.write('>' + header_2 + '\\n' + seq_2.upper() + '\\n')\n",
    "output_3.write('>' + header_3 + '\\n' + seq_3.replace('-','') + '\\n')\n",
    "output_1.close()\n",
    "output_2.close()\n",
    "output_3.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
